File: day25-go/part1/cucumber.go
package main

import "fmt"

type CucumberDirection int

const (
	East CucumberDirection = iota
	South
)

func (d CucumberDirection) getCoords() (int, int) {
	switch d {
	case East:
		return 1, 0
	case South:
		return 0, 1
	default:
		return 0, 0
	}
}

type Cucumber struct {
	row, col     int
	direction    CucumberDirection
	rawDirection rune
}

func NewCucumber(row, col int, direction rune) (*Cucumber, error) {
	var cucumberDirection CucumberDirection
	switch direction {
	case '>':
		cucumberDirection = East
	case 'v':
		cucumberDirection = South
	default:
		return nil, fmt.Errorf("Can't parse %c", direction)
	}

	return &Cucumber{
		row:          row,
		col:          col,
		direction:    cucumberDirection,
		rawDirection: direction,
	}, nil
}

func (c *Cucumber) nextStep(maxRow, maxCol int) (int, int) {
	switch c.direction {
	case South:
		return (c.row + 1) % maxRow, c.col
	case East:
		return c.row, (c.col + 1) % maxCol
	default:
		return c.row, c.col
	}
}

func (c *Cucumber) takeStep(maxRow, maxCol int) (int, int) {
	nextRow, nextCol := c.nextStep(maxRow, maxCol)
	c.row = nextRow
	c.col = nextCol
	return nextRow, nextCol
}

File: day25-rust/src/cucumber.rs
use std::fmt::Display;

#[derive(PartialEq, Eq)]
pub enum CucumberDirection {
    East,
    South,
    Unknown,
}

pub struct Cucumber {
    pub row: usize,
    pub col: usize,
    pub direction: CucumberDirection,
}

impl Cucumber {
    pub fn new(cucumber: &str, row_index: usize, col_index: usize) -> Option<Self> {
        let direction = match cucumber {
            ">" => CucumberDirection::East,
            "v" => CucumberDirection::South,
            _ => CucumberDirection::Unknown,
        };

        if let CucumberDirection::Unknown = direction {
            return None;
        }

        Some(Self {
            row: row_index,
            col: col_index,
            direction,
        })
    }

    pub fn take_step(&mut self, max_row: usize, max_col: usize) {
        let (next_row, next_col) = self.next_step(max_row, max_col);
        self.row = next_row;
        self.col = next_col;
    }

    pub fn next_step(&self, max_row: usize, max_col: usize) -> (usize, usize) {
        match self.direction {
            CucumberDirection::South => ((self.row + 1) % max_row, self.col),
            CucumberDirection::East => ((self.row, (self.col + 1) % max_col)),
            _ => (self.row, self.col),
        }
    }
}

impl Display for Cucumber {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let char_to_display = match self.direction {
            CucumberDirection::East => '>',
            CucumberDirection::South => 'v',
            CucumberDirection::Unknown => 'X',
        };
        write!(f, "{}", char_to_display)
    }
}
